air/src/execution_step/resolver/resolvable_impl.rs

```diff
@@ -0,0 +1,146 @@
+use crate::execution_step::ExecutionResult;
+use crate::JValue;
+use crate::LambdaAST;
+use crate::SecurityTetraplet;
+
+use air_parser::ast;
+
+use serde_json::json;
+use std::rc::Rc;
+
+/// Resolve value to called function arguments.
+impl Resolvable for ast::ImmutableValue<'_> {
+    fn resolve(&self, ctx: &ExecutionCtx<'_>) -> ExecutionResult<(JValue, RcSecurityTetraplets)> {
+        use ast::ImmutableValue::*;
+
+        match self {
+            InitPeerId => resolve_const(ctx.run_parameters.init_peer_id.as_str(), ctx),
+            LastError(error_accessor) => error_accessor.resolve(ctx),
+            Literal(value) => resolve_const(value.to_string(), ctx),
+            Timestamp => resolve_const(ctx.run_parameters.timestamp, ctx),
+            TTL => resolve_const(ctx.run_parameters.ttl, ctx),
+            Boolean(value) => resolve_const(*value, ctx),
+            Number(value) => resolve_const(value, ctx),
+            EmptyArray => resolve_const(json!([]), ctx),
+            Variable(variable) => variable.resolve(ctx),
+            VariableWithLambda(variable) => variable.resolve(ctx),
+        }
+    }
+}
+
+pub(crate) fn resolve_const(
+    arg: impl Into<JValue>,
+    ctx: &ExecutionCtx<'_>,
+) -> ExecutionResult<(JValue, RcSecurityTetraplets)> {
+    let jvalue = arg.into();
+    let tetraplet = SecurityTetraplet::literal_tetraplet(ctx.run_parameters.init_peer_id.as_ref());
+    let tetraplet = Rc::new(tetraplet);
+
+    Ok((jvalue, vec![tetraplet]))
+}
+
+impl Resolvable for JValue {
+    fn resolve(&self, ctx: &ExecutionCtx<'_>) -> ExecutionResult<(JValue, RcSecurityTetraplets)> {
+        let tetraplet = SecurityTetraplet::literal_tetraplet(ctx.run_parameters.init_peer_id.as_ref());
+        let tetraplet = Rc::new(tetraplet);
+
+        Ok((self.clone(), vec![tetraplet]))
+    }
+}
+
+impl Resolvable for Option<LambdaAST<'_>> {
+    fn resolve(&self, ctx: &ExecutionCtx<'_>) -> ExecutionResult<(JValue, RcSecurityTetraplets)> {
+        use crate::LastError;
+
+        let LastError { error, tetraplet } = ctx.last_error();
+
+        let jvalue = match self {
+            Some(error_accessor) => select_by_lambda_from_scalar(error.as_ref(), error_accessor, ctx)?.into_owned(),
+            None => error.as_ref().clone(),
+        };
+
+        let tetraplets = match tetraplet {
+            Some(tetraplet) => vec![tetraplet.clone()],
+            None => {
+                let tetraplet = SecurityTetraplet::literal_tetraplet(ctx.run_parameters.init_peer_id.as_ref());
+                let tetraplet = Rc::new(tetraplet);
+                vec![tetraplet]
+            }
+        };
+
+        Ok((jvalue, tetraplets))
+    }
+}
+
+impl Resolvable for ast::Scalar<'_> {
+    fn resolve(&self, ctx: &ExecutionCtx<'_>) -> ExecutionResult<(JValue, RcSecurityTetraplets)> {
+        let value = ctx.scalars.get_value(self.name)?.into_jvaluable();
+        let tetraplets = value.as_tetraplets();
+        Ok((value.into_jvalue(), tetraplets))
```

Comment:
Am I right that every time a value is fetched from a variable, it is cloned?
Looks like we can improve it (but that's for another PR, of course).

Reply:
Yep, this problem is derived from the previous resolving design, and I wanted to address it in the boxed value project.